### PR TITLE
[1075] SPIKE - Apply applications integration

### PR DESCRIPTION
**Pull request by AbigailMcP**
### Context

https://trello.com/c/5K3zuvwk/1075-2-days-apply-applications

### Changes proposed in this pull request

This PR adds a new (draft!) service `Trainees::CreateFromApply` that currently uses the example JSON from Apply to create a new trainee with as many of the fields filled in. I've included the JSON file in this PR for ease of review (i.e. you can run the service and test it yourself!)

I've also completed the following gap analysis:

#### Fields missing from API (but could be added):
| Our field                               | Closest match in Apply                                                                                                                                                                 | Any manipulation needed?                                                                                                                                                                 |
|-----------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| record_type (soon to be training_route) | course.program_type                                                                                                                                                                    | Map string values                                                                                                                                                                        |
| subject                                 | course.name (API already provides course_code)                                                                                                                                         | Map string values                                                                                                                                                                        |
| age_range                               | course.age_range                                                                                                                                                                       | Map string values                                                                                                                                                                        |
| provider                                | provider.name (API already provider provider_code)                                                                                                                                     | Check we have a matching provider                                                                                                                                                        |
| programme_end_date                      | course.course_length                                                                                                                                                                   | Add the course length to the programme_start_date                                                                                                                                        |
| gender                                  | application_form.equality_and_diversity["sex"]                                                                                                                                         | Already in matching format                                                                                                                                                               |
| diversity_disclosure                    | application_form.equality_and_diversity['ethnic_group'] \|\| application_form.equality_and_diversity['ethnic_background'] \|\| application_form.equality_and_diversity['disabilities'] | Deduce from the presence of these fields                                                                                                                                                 |
| disability_disclosure                   | application_form.equality_and_diversity['disabilities']                                                                                                                                | If empty, then disability_disclosure = :no_disability. If "Prefer not to say", then disability_disclosure = :not_provided If disabilities listed, then disability_disclosure = :disabled |
| ethnic_group                            | application_form.equality_and_diversity['ethnic_group']                                                                                                                                | Already in matching format                                                                                                                                                               |
| ethnic_background                       | application_form.equality_and_diversity['ethnic_background']                                                                                                                           | Already in matching format                                                                                                                                                               |
| additional_ethnic_background            | application_form.equality_and_diversity['ethnic_background']                                                                                                                           | (If not included in our enum)                                                                                                                                                            |
| non_uk_degree            | application_qualifications.comparable_uk_degree                                                                                                                           | Map string values                                                                                                                                                            |

Fields not in Apply:
- dttp_id
- placement_assignment_dttp_id
- middle_names

#### Not included in this PR (yet?)
- An additional field on trainee - `source`. This would be set to `apply` within this service, so that we can distinguish apply trainees at a later date.

### Guidance to review

- Start up a rails console `rails c`
- Run `Trainees::CreateFromApply.call`
- Check that a trainee is created with the expected attributes

### CODE REVIEW

**Comment by AbigailMcP** [comment]
[[2 days] Apply applications](https://trello.com/c/5K3zuvwk/1075-2-days-apply-applications)

**Comment by defong** [comment]
> #### Fields missing from API (but could be added):
| Our field                               | Closest match in Apply                                                                                                                                                                 | Any manipulation needed?                                                                                                                                                                 |
|-----------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| record_type (soon to be training_route) | course.program_type                                                                                                                                                                    | Map string values                                                                                                                                                                        |
| subject                                 | course.name (API already provides course_code)                                                                                                                                         | Map string values                                                                                                                                                                        |
| age_range                               | course.age_range                                                                                                                                                                       | Map string values                                                                                                                                                                        |
| provider                                | provider.name (API already provider provider_code)                                                                                                                                     | Check we have a matching provider                                                                                                                                                        |
| programme_end_date                      | course.course_length                                                                                                                                                                   | Add the course length to the programme_start_date                                                                                                                                        |                                                                                                                                     

These five items, _**source of truth**_, derives from https://github.com/DFE-Digital/teacher-training-api/blob/1cff52b7d5283c2fe2d9ba4f9094ebd0524c4a54/app/models/course.rb

with varying difficulties and issues, for example
- course length vs date, relative value,
- course name vs subject, marketing name
- provider name vs provider, marketing name

File: app/services/trainees/create_from_apply.rb [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+    def trainee
+      @trainee ||= Trainee.new(mapped_attributes)
+    end
+
+    def mapped_attributes
+      {
+        # TODO: Use actual provider, if found.
+        provider: Provider.find(1),
+        trainee_id: raw_trainee["id"],
```
Comment (defong) [inline app/services/trainees/create_from_apply.rb:36]: a bit annoying, but it is what it is...

File: app/services/trainees/create_from_apply.rb [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+      Date.new(year, month)
+    end
+
+    def address
+      raw_contact_details["country"] != "GB" ? international_address : uk_address
+    end
+
+    def international_address
+      {
+        international_address: raw_contact_details["address_line1"],
```
Comment (defong) [inline app/services/trainees/create_from_apply.rb:58]: I wonder what the none uk contact details shape looks like?

File: lib/example_apply_response.json [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+          "hesa_degtype": null,
+          "hesa_degsbj": null,
+          "hesa_degclss": null,
+          "hesa_degest": null,
+          "hesa_degctry": null,
+          "hesa_degstdt": null,
+          "hesa_degenddt": null
+        }
+      ],
+      "degrees": [
```
Comment (defong) [inline lib/example_apply_response.json:128]: looks like the same/similar degrees as ours

**Review by defong — COMMENTED** [review]
looks good

File: lib/example_apply_response.json (hunk 3 above)
Comment (AbigailMcP) [reply lib/example_apply_response.json:128]: ahhh degrees! i knew there was something else I wanted to check - thanks!

File: app/services/trainees/create_from_apply.rb (hunk 2 above)
Comment (duncanjbrown) [reply app/services/trainees/create_from_apply.rb:63]: This has now changed — it used to be all in one line, now it's split across four. We corrected all the old data to be that shape too

File: app/services/trainees/create_from_apply.rb (hunk 1 above)
Comment (duncanjbrown) [reply app/services/trainees/create_from_apply.rb:36]: What's the matter here @defong ? Can we improve on the Apply side?

Comment (defong) [reply app/services/trainees/create_from_apply.rb:36]: `trainee_id` nothing to do with Apply, we just have an identifier or field that is solely of questionable nature. 

Unless on Apply, the provider additional gives the candidate another symbolic name, such as `cohort 2021`, to remove the need to refer it by PII.

**Review by defong — COMMENTED** [review]
Degrees stuff looking nice

File: app/services/trainees/create_from_apply.rb (hunk 1 above)
Comment (edwardhorsford) [reply app/services/trainees/create_from_apply.rb:36]: We might want to rename our field anyway to make it clear it's the provider's own internal id for the trainee.

**Comment by AgaDufrat** [comment]
👋 Hi team, We [started drafting the presenters](https://github.com/DFE-Digital/apply-for-teacher-training/compare/3499-spike-bare-bones-api-for-register). I have a few questions about the attributes
1. `non_uk_degree` – Apply's `application_qualifications.comparable_uk_degree` can have the following values:
- bachelor_ordinary_degree
- masters_degree
- doctor_of_philosophy
- post_doctoral_award
- postgraduate_certificate_or_diploma
- bachelor_honours_degree

Is this what you are after?

2. `programme_end_date` I don't think it will be possible to calculate the end date because `course_length` is a string, with typical values e.g. “OneYear”, “TwoYears” but there is some free text data as well. Are you ok with us providing `course_length` like that? We could also present it in more human readable format using [this](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/app/services/display_course_length.rb).

**Comment by AbigailMcP** [comment]
> 👋 Hi team, We [started drafting the presenters](https://github.com/DFE-Digital/apply-for-teacher-training/compare/3499-spike-bare-bones-api-for-register). I have a few questions about the attributes
> 
> 1. `non_uk_degree` – Apply's `application_qualifications.comparable_uk_degree` can have the following values:
> 
> * bachelor_ordinary_degree
> * masters_degree
> * doctor_of_philosophy
> * post_doctoral_award
> * postgraduate_certificate_or_diploma
> * bachelor_honours_degree
> 
> Is this what you are after?

Hi @AgaDufrat ! Thanks for this.

Re `non_uk_degree`, I believe that's what we're after. In Register, the possible choices via the interface for `non_uk_degree` are:

- "Bachelor degree"
- "Ordinary bachelor degree"
- "Bachelor degree with honours"
- "Postgraduate certificate or postgraduate diploma"
- "Master’s degree or Integrated Master’s degree"
- "non_naric"

(saved as a string in the DB)

So it will be a case of us matching up what Apply sends to these values 👍

**Pull request closed without merging**